network/firewall_test.go

```diff
@@ -0,0 +1,196 @@
+)
+
+func MockNetworkClient() *network.NetworkClient {
+	return &network.NetworkClient{
+		Client: testutils.NewMockClient(testutils.MockClientInput{
+			AccountName: accountUrl,
+		}),
+	}
+}
+
+func TestListRuleMachines(t *testing.T) {
+	networkClient := MockNetworkClient()
+
+	do := func(ctx context.Context, nc *network.NetworkClient) ([]*network.Machine, error) {
+		defer testutils.DeactivateClient()
+
+		machines, err := nc.Firewall().ListRuleMachines(ctx, &network.ListRuleMachinesInput{
+			ID: "38de17c4-39e8-48c7-a168-0f58083de860",
+		})
+		if err != nil {
+			return nil, err
+		}
+		return machines, nil
+	}
+
+	t.Run("successful", func(t *testing.T) {
+		testutils.RegisterResponder("GET", fmt.Sprintf("/%s/fwrules/%s/machines", accountUrl, "38de17c4-39e8-48c7-a168-0f58083de860"), listRuleMachinesSuccess)
+
+		resp, err := do(context.Background(), networkClient)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if resp == nil {
+			t.Fatalf("Expected an output but got nil")
+		}
+	})
+
+	t.Run("eof", func(t *testing.T) {
+		testutils.RegisterResponder("GET", fmt.Sprintf("/%s/fwrules/%s/machines", accountUrl, "38de17c4-39e8-48c7-a168-0f58083de860"), listRuleMachinesEmpty)
+
+		_, err := do(context.Background(), networkClient)
+		if err == nil {
+			t.Fatal(err)
+		}
+
+		if !strings.Contains(err.Error(), "EOF") {
+			t.Errorf("expected error to contain EOF: found %s", err)
+		}
+	})
+
+	t.Run("bad_decode", func(t *testing.T) {
+		testutils.RegisterResponder("GET", fmt.Sprintf("/%s/fwrules/%s/machines", accountUrl, "38de17c4-39e8-48c7-a168-0f58083de860"), listRuleMachinesBadDecode)
+
+		_, err := do(context.Background(), networkClient)
+		if err == nil {
+			t.Fatal(err)
+		}
+
+		if !strings.Contains(err.Error(), "invalid character") {
+			t.Errorf("expected decode to fail: found %s", err)
+		}
+	})
+
+	t.Run("error", func(t *testing.T) {
+		testutils.RegisterResponder("GET", fmt.Sprintf("/%s/fwrules/%s/machines", accountUrl, "38de17c4-39e8-48c7-a168-0f58083de860"), listRuleMachinesError)
+
+		resp, err := do(context.Background(), networkClient)
+		if err == nil {
+			t.Fatal(err)
+		}
+		if resp != nil {
+			t.Error("expected resp to be nil")
+		}
+
+		if !strings.Contains(err.Error(), "Error executing ListRuleMachines request:") {
+			t.Errorf("expected error to equal testError: found %s", err)
+		}
+	})
+}
+
+func listRuleMachinesEmpty(req *http.Request) (*http.Response, error) {
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+	return &http.Response{
+		StatusCode: 200,
+		Header:     header,
+		Body:       ioutil.NopCloser(strings.NewReader("")),
+	}, nil
+}
+
+func listRuleMachinesSuccess(req *http.Request) (*http.Response, error) {
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	body := strings.NewReader(`[
+	{
+    "id": "b6979942-7d5d-4fe6-a2ec-b812e950625a",
+    "name": "test",
+    "type": "smartmachine",
```

Comment:
Is there anything different between a Docker container or native machine? Are we testing that?

Reply:
We are not testing that - the API itself should take care of that - there are no docs about what is supported here. We will just return a list of machines as expected

Reply:
👍 Pretty much all I wanted to know.